JEP.md

```diff
@@ -0,0 +1,293 @@
+2. As a Jupyter user, I need to opt-in to event collection, so that I can choose what data about my usage is gathered.
+3. As a Jupyter extension developer, I can publish custom events using a well defined interface, from the browser as well as the server.
+4. As a Jupyter operator, I can configure events sinks for my deployment, so that I have control over where events are sent to.
+5. As a Jupyter operator, I can write a custom event sink for my deployment, so that I can control where event data is stored.
+6. As a Jupyter operator, I can decide what kinds and levels of events are going to be collected and stored.
+7. As a Jupyter operator, I can configure opt-in or opt-out for the users in my deployment.
+8. As an analyst,  I can clearly tell what the fields in the data represent so that I can interpret and make decisions from that data.
+
+## Proposed Enhancement
+
+The enhancement spans two categories - general concepts applicable across all Jupyter components, and the implementation for each concept for each Jupyter component.
+
+**General concepts:**
+
+* Event schema: The schema for a given event which is respected by all the Jupyter components across the event's lifecycle.
+* Event sinks: The "backends" which handle publishing the event to its final storage location.
+
+**Jupyter components:**
+
+* Jupyter server
+* JupyterLab
+* JupyterHub
+* Jupyter Classic
+
+## Detailed Explanation
+
+### Event Schema
+
+Each event is published conforming a given event schema and this is the contract followed by each step in the event lifecycle. These are referred to by all publishers regardless of whether the event comes from the server or the browser.
+
+Example event schema
+
+```json
+{
+    "name": "org.jupyter.kernel_lifecycle_event",
+    "version": 1,
+    "title": "Event emitted when a kernel lifecycle state changes",
+    "description": "Records each invocation of any command in JupyterLab",
+    "type": "object",
+    "args": {
+        "kernel_name": {
+            "type": "string",
+            "description": "Name of the kernel"
+        },
+        "state": {
+            "type": "string",
+            "description": "The new state of the kernel"
+        }
+    }
+}
+```
+
+Schema validation is done in the core telemetry framework that routes events from publishers to sinks, so that each event sink does not need to implement schema validation.
+
+**Open Questions**:
+
+* How do organizations deploy custom schemas?
+* Where are the schemas for public events stored?
+
+### Event Sinks
+
+Event sinks are the backends where events published to.  Event sinks can be configured from the browser as well as the server. Custom sinks can be implemented by extending the interface for the given Jupyter component. 
+See [Python interface](#python-event-sink-interface) and the [JupyterLab interface](#jupyterlab-event-sink-interface)
+
+### Jupyter Server
```

Comment:
Do we expect the telemetry library to be used outside the Jupyter server or  server extensions ? Its really tied to the the Jupyter server's configuration

Reply:
No, I don't expect telemetry to be used outside of jupyter server; but I could see people wanting a jupyter server *without telemetry*? 

That's why I was thinking that making telemetry a server extension might make the most sense.

But you're right. Telemetry has to be hardcoded into and configured directly by the server. So this means adding a new "service" to our core [REST API](http://petstore.swagger.io/?url=https://raw.githubusercontent.com/jupyter/notebook/master/notebook/services/api/api.yaml).

Reply:
I think what I meant by server isn't clear - What I mean by server is any "non-frontend" component, i.e., the actual Jupyter web-server (whether its from the classic notebook or the new jupyter_server). So this includes the Python libs, the Jupyter config files, and the REST API. I'm open to change the name of the heading based on this.

Reply:
I see. We could rename this section to "Telemetry components" or something like that. 

I still think there's this question of whether telemetry becomes a core service, and therefore, a dependency of the jupyter_server.

Reply:
Based on my understanding in https://github.com/jupyterlab/jupyterlab-telemetry/blob/master/design.md#design-proposal, most of the logic should be in a core Python library that can be used in jupyter server, jupyterhub, binderhub, hubtraf, etc.

Reply:
@yuvipanda Oh, right. This isn't a service at all.  

The "telemetry" library is a Python API (with a matching JS API) that offers an "EventLogger" configurable object. We'll add this object as an attribute to various server-type applications (i.e. jupyter_server, jupyterhub, binderhub, jupyterlab). If it's not configured (i.e. given event sinks and white-listed events), that application won't emit events. 

To collect events, we're letting administrators add "EventHandlers" (using Python logging lingo) to the EventLogger through Jupyter's configuration system. Those handlers mediate where+how events are sent. 

We'll provide some base EventHandlers (e.g. HTTPEventHandler or FileEventHandler) and EventSinks that can be extended to link our EventLogger to any event collecting service. 

We'll **also** provide a REST API that can be added to any application to *make it an event sink*. For example, one could extend a running notebook/jupyter server with this REST endpoint and that server will begin collecting events. That way, you wouldn't have to launch some external service to collect events.  

Am I thinking about this correctly?

Reply:
Yes, when I said "Jupyter Server" , I meant the server side components of Jupyter rather than the literal [jupyter_server](https://github.com/jupyter/jupyter_server)

The core logic is still in Python library but its a bit more - We have defined interfaces for event publishers, event sinks and a mediator which can route to event sinks based on the configuration.

> We'll also provide a REST API that can be added to any applicatio

I don't know if we need to add it to **any** application, but we'll add it to the Notebook server (Classic) and new [jupyter_server](https://github.com/jupyter/jupyter_server)